crates/jxl-frame/src/lib.rs

```diff
@@ -97,6 +97,25 @@ impl<'a> Bundle<&'a ImageHeader> for Frame<'a> {
             return Err(jxl_bitstream::Error::ValidationFailed("lf_level out of range").into());
         }
 
+
+        for (ec_upsampling, ec_info) in header
+            .ec_upsampling
+            .iter()
+            .zip(image_header.metadata.ec_info.iter())
+        {
+            let ec_up_cumulative = ec_upsampling << ec_info.dim_shift;
+            if ec_up_cumulative < header.upsampling {
+                return Err(jxl_bitstream::Error::ValidationFailed(
+                    "EC upsampling < color upsampling, which is invalid").into()
+                );
+            }
+            if ec_up_cumulative > 8 {
```

Comment:
As far as I got it right, it's the `dim_shift` that makes the actual encoded block dimension smaller (to `group_dim >> dim_shift`), so only `dim_shift` should be checked here. The spec says that `1 << dim_shift` does not exceed `group_dim`, so it makes sense; otherwise the actual dimension would be smaller than 1.

The required assertion is that `(1 << dim_shift) <= group_dim`, where `group_dim = 1 << (7 + group_size_shift)`, so we need to check whether `ec.dim_shift <= 7 + frame_header.group_size_shift` holds.